src/common/w25x.h

```diff
@@ -35,12 +35,27 @@ inline constexpr uint32_t w25x_fs_start_address = W25X_FS_START_ADDRESS;
 /// This has to be called after the underlying SPI has been initialized
 /// and assigned using w25x_spi_assign.
 ///
-/// @param init_dma Dynamically allocate resources for DMA transfers
-///        DMA and RTOS facilities are not used if false,
-///        it has no effect if it was already allocated by previous call
+/// When w25x is initialized when the scheduler is running all its
+/// interface function must be called in task context only. (standard usage)
+///
+/// When w25x is initialized when the scheduler is NOT running all its
+/// interface function can be called in any context but are not reentrant.
+///
+/// w25x_init can be called repeatedly in different contexts
+/// to switch between those two modes. It can be called only once with
+/// running scheduler as this creates resources which are never released.
+/// If w25x is reinitialized during DMA transfer it is aborted. If some
+/// data is already transfered to the chip at that point those data are
+/// written gracefully. If erase operation is ongoing it is completed
+/// during reinitialization.
+///
+/// Worst case runtime is 100 seconds if called just after chip erase
+/// operation has been started. Worst case runtime is 200 seconds for
+/// maliciously crafted w25x responses.
```

Comment:
Yes those are 200 seconds. so 3.3 minutes. You are right.

We might reduce maximum timeout from 100 s to 2 s and so worst case `w25x_init()` runtime to 4 seconds if we remove chip erase support. Doing this would decrease our guaranteed erase throughput from 82 to 32 kB/s. But at this moment we don't use chip erase meaningfully. We have it only accessible through debug menu. Only meaningful usage of chip erase would be to speedup firmware update bootstrap process. But at this moment w25x is not even bottleneck of resource bootstrap. Currently 500kB resource bootstrap takes 12 minutes on my printer and of this process 11 minutes takes reading file from USB and 1 minute to write it to w25x. 

If we have 8 MB of resources in future, writing those to w25x would take:
| Time to update 8MB resources | TYP | MAX | unit |
| ------------- | ------------- | ------------- | ------------- |
| with chip erase available | 0.5 | 3.3 | minutes |
| with chip erase removed using 64KB erase | 0.5  | 6 | minutes |
| using only 4KB erase as currently | 1.75 | 15.3 | minutes |
| using only 4KB erase and inefficient USB access as currently | 3.2 |  | hours |

| Description | TYP | MAX | Unit | Throughput guaranteed .. typical |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| read |  | 50| MHz | 6.25 MB/s |
| fast read, write | | 133 | MHz | 16.6 MB/s |
| Page Program Time | 0.4 | 3 | ms | 83 .. 625 kB/s |
| Sector Erase Time (4KB) |  45 | 400 | ms |  10 .. 88 kB/s |
| Block Erase Time (32KB) |  120 | 1 600 | ms | 20 .. 266 kB/s |
| Block Erase Time (64KB) |  150 | 2 000 |  ms | 32 .. 426 kB/s |
| Chip Erase Time | 20 | 100 | s | 82 .. 409 kB/s |